core/jvm/src/main/scala/zio/internal/Blocking.scala

```diff
@@ -26,4 +42,13 @@ object Blocking {
 
       threadPool
     }
+
+  /**
+   * Signals to the scheduler that the current thread is about to block. This
+   * method is a no-op except when the current thread is a ZScheduler.Worker. In
+   * that case, the worker is marked as "blocking" and a new worker is spawned
+   * to replace it.
+   */
+  private[zio] final def signalBlocking(): Unit =
+    ZScheduler.markCurrentWorkerAsBlocking()
```

Comment:
so this is a ZScheduler specific capability? `scala.concurrent.blocking` relies on a thread-local to communicate back to the effective execution-context

Reply:
Yes. It's only a problem in that case, because the `ZScheduler.Worker` will be creating a fiber that will also require another Worker to run it, and thus ending up in a deadlock. This work isn't about optimizing for running blocking code, it's about preventing deadlocks due to having reentrance

Reply:
just saying, the runtime may be configured with a different executor which can exhibit the same deadlock prone behavior... I guess it's ok to support this only in ZSchduler atm, we can always generalize it later on